### PR TITLE
Implement LatestByIndexed on cpp side

### PR DESCRIPTION
**Pull request by eugenels**

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:disappointed: **fail** : 67 / 137 (48.91%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/BitmapIndexUtilsNative.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexFwdNullReader.java|0|6|00.00%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexBwdNullReader.java|0|6|00.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursor.java|20|48|41.67%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedRecordCursor.java|20|39|51.28%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByArguments.java|11|19|57.89%|
|:large_blue_circle:|io/questdb/std/DirectLongList.java|6|8|75.00%|
|:large_blue_circle:|io/questdb/cairo/AbstractIndexReader.java|10|10|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:disappointed: **fail** : 67 / 137 (48.91%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/BitmapIndexUtilsNative.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexFwdNullReader.java|0|6|00.00%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexBwdNullReader.java|0|6|00.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursor.java|20|48|41.67%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedRecordCursor.java|20|39|51.28%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByArguments.java|11|19|57.89%|
|:large_blue_circle:|io/questdb/std/DirectLongList.java|6|8|75.00%|
|:large_blue_circle:|io/questdb/cairo/AbstractIndexReader.java|10|10|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:disappointed: **fail** : 67 / 137 (48.91%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/BitmapIndexUtilsNative.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexFwdNullReader.java|0|6|00.00%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexBwdNullReader.java|0|6|00.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursor.java|20|48|41.67%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedRecordCursor.java|20|39|51.28%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByArguments.java|11|19|57.89%|
|:large_blue_circle:|io/questdb/std/DirectLongList.java|6|8|75.00%|
|:large_blue_circle:|io/questdb/cairo/AbstractIndexReader.java|10|10|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 119 / 137 (86.86%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/BitmapIndexUtilsNative.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexFwdNullReader.java|0|6|00.00%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexBwdNullReader.java|0|6|00.00%|
|:large_blue_circle:|io/questdb/std/DirectLongList.java|6|8|75.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByArguments.java|16|19|84.21%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursor.java|48|48|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedRecordCursor.java|39|39|100.00%|
|:large_blue_circle:|io/questdb/cairo/AbstractIndexReader.java|10|10|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 116 / 134 (86.57%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/BitmapIndexUtilsNative.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexFwdNullReader.java|0|6|00.00%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexBwdNullReader.java|0|6|00.00%|
|:large_blue_circle:|io/questdb/std/DirectLongList.java|6|8|75.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByArguments.java|16|19|84.21%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursor.java|48|48|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedRecordCursor.java|39|39|100.00%|
|:large_blue_circle:|io/questdb/cairo/AbstractIndexReader.java|7|7|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 116 / 134 (86.57%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/BitmapIndexUtilsNative.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexFwdNullReader.java|0|6|00.00%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexBwdNullReader.java|0|6|00.00%|
|:large_blue_circle:|io/questdb/std/DirectLongList.java|6|8|75.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByArguments.java|16|19|84.21%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursor.java|48|48|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedRecordCursor.java|39|39|100.00%|
|:large_blue_circle:|io/questdb/cairo/AbstractIndexReader.java|7|7|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 116 / 134 (86.57%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/BitmapIndexUtilsNative.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexFwdNullReader.java|0|6|00.00%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexBwdNullReader.java|0|6|00.00%|
|:large_blue_circle:|io/questdb/std/DirectLongList.java|6|8|75.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByArguments.java|16|19|84.21%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursor.java|48|48|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedRecordCursor.java|39|39|100.00%|
|:large_blue_circle:|io/questdb/cairo/AbstractIndexReader.java|7|7|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 92 / 143 (64.34%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursor.java|20|48|41.67%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedRecordCursor.java|20|39|51.28%|
|:large_blue_circle:|io/questdb/std/BitmapIndexUtilsNative.java|8|10|80.00%|
|:large_blue_circle:|io/questdb/std/DirectLongList.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByArguments.java|18|19|94.74%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexFwdNullReader.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexBwdNullReader.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cairo/AbstractIndexReader.java|7|7|100.00%|

File: core/src/main/c/share/bitmap_index_utils.cpp [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+                    row_count += 1;
+                    update_range = false;
+                }
+            }
+        }
+
+        // unindexed nulls case
+        if (k == 0 && unindexed_null_count > 0) {
+            if (rows[k] <= 0 && unindexed_null_count - 1 >= min_value) {
+                rows[k] = ((int64_t) partition_index << 44) + unindexed_null_count;
```
Comment (bluestreak01) [inline core/src/main/c/share/bitmap_index_utils.cpp:118]: we should perhaps refactor a helper method here, which creates rowid, perhaps an abstract type ?

File: core/src/main/c/share/bitmap_index_utils.cpp [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+            }
+
+            auto values = reinterpret_cast<const int64_t *>(values_memory);
+            auto res = seekValueBlockRTL(value_count, last_vblock_offset, values, max_value,
+                                         vblock_capacity_mask);
+            value_count = res.first;
+            last_vblock_offset = res.second;
+            if (value_count > 0) {
+                uint64_t cell_index = --value_count & vblock_capacity_mask;
+                int64_t local_row_id = values[last_vblock_offset / 8 + cell_index];
```
Comment (bluestreak01) [inline core/src/main/c/share/bitmap_index_utils.cpp:106]: is there a way to get c++ do this arithmetic ?

**Review by bluestreak01 — CHANGES_REQUESTED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:disappointed: **fail** : 0 / 147 (00.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/BitmapIndexUtilsNative.java|0|10|00.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursor.java|0|48|00.00%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexFwdNullReader.java|0|6|00.00%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexBwdNullReader.java|0|6|00.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedRecordCursor.java|0|39|00.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByArguments.java|0|19|00.00%|
|:large_blue_circle:|io/questdb/cairo/AbstractIndexReader.java|0|7|00.00%|
|:large_blue_circle:|io/questdb/std/DirectLongList.java|0|12|00.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 146 / 147 (99.32%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByArguments.java|18|19|94.74%|
|:large_blue_circle:|io/questdb/std/BitmapIndexUtilsNative.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursor.java|48|48|100.00%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexFwdNullReader.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexBwdNullReader.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedRecordCursor.java|39|39|100.00%|
|:large_blue_circle:|io/questdb/cairo/AbstractIndexReader.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/std/DirectLongList.java|12|12|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 146 / 147 (99.32%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByArguments.java|18|19|94.74%|
|:large_blue_circle:|io/questdb/std/BitmapIndexUtilsNative.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursor.java|48|48|100.00%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexFwdNullReader.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexBwdNullReader.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedRecordCursor.java|39|39|100.00%|
|:large_blue_circle:|io/questdb/cairo/AbstractIndexReader.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/std/DirectLongList.java|12|12|100.00%|

**Review by bluestreak01 — APPROVED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 146 / 147 (99.32%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByArguments.java|18|19|94.74%|
|:large_blue_circle:|io/questdb/std/BitmapIndexUtilsNative.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursor.java|48|48|100.00%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexFwdNullReader.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexBwdNullReader.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedRecordCursor.java|39|39|100.00%|
|:large_blue_circle:|io/questdb/cairo/AbstractIndexReader.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/std/DirectLongList.java|12|12|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 146 / 147 (99.32%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByArguments.java|18|19|94.74%|
|:large_blue_circle:|io/questdb/std/BitmapIndexUtilsNative.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursor.java|48|48|100.00%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexFwdNullReader.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexBwdNullReader.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedRecordCursor.java|39|39|100.00%|
|:large_blue_circle:|io/questdb/cairo/AbstractIndexReader.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/std/DirectLongList.java|12|12|100.00%|

**Pull request merged**